### PR TITLE
SHE: auto-detect TLS and create secure etcd client

### PR DESCRIPTION
**Pull request by hongchaodeng**
step 3 of https://github.com/kubernetes-incubator/bootkube/issues/582

### CODE REVIEW

File: cmd/bootkube/recover.go [hunk 1]
```diff
@@ -87,7 +87,7 @@ func runCmdRecover(cmd *cobra.Command, args []string) error {
 
 		bootkube.UserOutput("Waiting for etcd server to start...\n")
 
-		err = etcdutil.WaitClusterReady(recovery.RecoveryEtcdClientAddr)
+		err = etcdutil.WaitEtcdClusterReady(recovery.RecoveryEtcdClientAddr, nil)
```
Comment (xiang90) [inline cmd/bootkube/recover.go]: Remove Etcd. this is already in etcdutil pkg.

Comment (hongchaodeng) [inline cmd/bootkube/recover.go]: fixed

**Comment by xiang90** [comment]
the vendoring change is massive. etcd tlsutil should be a small pkg, no?

**Comment by hongchaodeng** [comment]
There is a `make vendor` command in bootkube.
I just run that.
If there are any other ways to update vendor, let me know.

**Comment by hongchaodeng** [comment]
https://github.com/kubernetes-incubator/bootkube/pull/588 should reduce the dep change of this PR

**Comment by hongchaodeng** [comment]
Rebased to reduce dep change.

Good to review now.

**Comment by hongchaodeng** [comment]
coreosbot run e2e

File: pkg/bootkube/bootkube.go [hunk 2]
```diff
@@ -86,15 +88,39 @@ func (b *bootkube) Run() error {
...
+		err = etcdutil.Migrate(kubeConfig, tlsConfig, svcPath, tprPath)
+		if err != nil {
 			return err
 		}
 	}
 
 	return nil
 }
 
+func (b *bootkube) makeEtcdTLS() (*tls.Config, error) {
```
Comment (diegs) [inline pkg/bootkube/bootkube.go]: move to etcdutil?

File: glide.lock [hunk 3]
```diff
@@ -34,6 +35,12 @@ imports:
   - key
   - oauth2
   - oidc
+- name: github.com/coreos/go-systemd
```
Comment (diegs) [inline glide.lock]: where is this dep used?

Comment (hongchaodeng) [inline glide.lock]: It's brought along with `coreos/etcd/pkg/transport`

**Comment by hongchaodeng** [comment]
fixed

File: cmd/bootkube/recover.go [hunk 4]
```diff
@@ -87,7 +87,7 @@ func runCmdRecover(cmd *cobra.Command, args []string) error {
 
 		bootkube.UserOutput("Waiting for etcd server to start...\n")
 
-		err = etcdutil.WaitClusterReady(recovery.RecoveryEtcdClientAddr)
+		err = etcdutil.WaitClusterReady(recovery.RecoveryEtcdClientAddr, nil)
```
Comment (diegs) [inline cmd/bootkube/recover.go]: Can you add a TODO or something to add self-hosted etcd+TLS support?

File: pkg/bootkube/bootkube.go [hunk 5]
```diff
@@ -86,8 +87,23 @@ func (b *bootkube) Run() error {
 	}
 
 	if selfHostedEtcd {
+		useEtcdTLS, err := detectEtcdTLS(b.assetDir)
+		if err != nil {
+			return err
+		}
+		var tlsConfig *tls.Config
```
Comment (diegs) [inline pkg/bootkube/bootkube.go]: Well, I meant you could move all of this to etcdutil too. Maybe just pass the assetDir into Migrate() and you can construct the other two paths inside that method as well.

**Comment by hongchaodeng** [comment]
fixed

**Comment by hongchaodeng** [comment]
coreosbot run e2e

**Pull request merged**